File: JavaScript/containsduplicate.js
// https://leetcode.com/problems/contains-duplicate/



// Hash Map Approach
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * @param {number[]} nums
 * @return {boolean}
 */
var containsDuplicate = function(nums) {
    let dict = {};
    for (let num of nums) {
        if (dict[num] >= 1) {
            return true;
        }

        dict[num] = 1;
    } 
    return false;
};

// Hash Set Approach
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * @param {number[]} nums
 * @return {boolean}
 */
var containsDuplicate = function(nums) {
    let set = new Set();
    for (let num of nums) {
        if (set.has(num)) {
            return true;
        }
        set.add(num);
    }
    return false;
};

console.log(containsDuplicate([1, 2, 4, 4]))
console.log(containsDuplicate([1, 2, 3]))

File: Python/arrays_and_hashing/containsduplicate.py
# https://leetcode.com/problems/contains-duplicate

from typing import List

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        hashset = set()
        for num in nums:
            if num in hashset:
                return True
            hashset.add(num)
        return False

solution = Solution()
print(solution.containsDuplicate([1, 2, 3, 4]))
print(solution.containsDuplicate([1, 2, 3, 2]))